src/sparse/KokkosSparse_spadd.hpp

```diff
@@ -489,48 +499,49 @@ struct SortedNumericSumFunctor {
         Cvalues(Cvalues_),
         alpha(alpha_),
         beta(beta_) {}
-  KOKKOS_INLINE_FUNCTION void operator()(const ordinal_type i) const {
-    size_type CrowStart = Crowptrs(i);
-    size_type ArowStart = Arowptrs(i);
-    size_type ArowEnd   = Arowptrs(i + 1);
-    size_type Arowlen   = ArowEnd - ArowStart;
-    size_type BrowStart = Browptrs(i);
-    size_type BrowEnd   = Browptrs(i + 1);
-    size_type Browlen   = BrowEnd - BrowStart;
-    size_type ai        = 0;
-    size_type bi        = 0;
-    size_type ci        = 0;
-    // add in A entries, while setting C colinds
-    while (ai < Arowlen && bi < Browlen) {
-      ordinal_type Acol = Acolinds(ArowStart + ai);
-      ordinal_type Bcol = Bcolinds(BrowStart + bi);
-      if (Acol <= Bcol) {
-        Ccolinds(CrowStart + ci) = Acol;
-        Cvalues(CrowStart + ci) += alpha * Avalues(ArowStart + ai);
+
+  KOKKOS_INLINE_FUNCTION void operator()(const ordinal_type i) const
+  {
+    // count the union of nonzeros in Arow and Brow
+    size_type ai         = 0;
+    size_type bi         = 0;
+    size_type Arowstart  = Arowptrs(i);
+    size_type Arowlen    = Arowptrs(i + 1) - Arowstart;
+    size_type Browstart  = Browptrs(i);
+    size_type Browlen    = Browptrs(i + 1) - Browstart;
+    ordinal_type Acol = (Arowlen == 0) ? ORDINAL_MAX : Acolinds(Arowstart);
+    ordinal_type Bcol = (Browlen == 0) ? ORDINAL_MAX : Bcolinds(Browstart);
+    size_type Coffset = Crowptrs(i);
+    while (Acol != ORDINAL_MAX || Bcol != ORDINAL_MAX)
+    {
+      ordinal_type Ccol = (Acol < Bcol) ? Acol : Bcol;
+      //Eat all entries in both A and B which have this column
+      //This also results in Acol/Bcol being updated to following entries for next loop iter
+      CscalarT accum = {0};
```

Comment:
I would have thought that `CscalarT accumulation = Kokkos::ArithTraits<CscalarT>::zero();` is the correct way to initialize `accum` to handle various complex types and Stokhos defines specializations for UQ_PCE and MP_VECTORS. Does this version work OK for exotic types?

Reply:
@lucbv Good point, I agree that I should use ArithTraits here.